File: server/services/deepseek.js
const axios = require('axios')
require('dotenv').config()

const API_URL = 'https://openrouter.ai/api/v1/chat/completions'
const API_KEY = process.env.OPENROUTER_API_KEY

async function generateContent(topic, keywords, tone, type) {
  const prompt = `You are a professional content writer. Create a ${type} about "${topic}".
Keywords to include: ${keywords}
Tone: ${tone}
Requirements:
- SEO-optimized with natural keyword placement
- Well-structured with headings and sections
- Engaging and informative
- Appropriate length for a ${type}
Write the content now:`

  const response = await axios.post(API_URL, {
    model: 'deepseek/deepseek-chat-v3-0324',
    messages: [{ role: 'user', content: prompt }],
  }, {
    headers: {
      'Authorization': `Bearer ${API_KEY}`,
      'Content-Type': 'application/json',
    },
  })

  return response.data.choices[0].message.content
}

module.exports = { generateContent }

File: go-service/services/deepseek.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

const openRouterURL = "https://openrouter.ai/api/v1/chat/completions"

type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatRequest struct {
	Model    string        `json:"model"`
	Messages []chatMessage `json:"messages"`
	Stream   bool          `json:"stream,omitempty"`
}

type chatChoice struct {
	Message chatMessage `json:"message"`
}

type chatResponse struct {
	Choices []chatChoice `json:"choices"`
}

func GenerateContent(topic, keywords, tone, contentType string) (string, error) {
	prompt := fmt.Sprintf(`You are a professional content writer. Create a %s about "%s".
Keywords to include: %s
Tone: %s
Requirements:
- SEO-optimized with natural keyword placement
- Well-structured with headings and sections
- Engaging and informative
- Appropriate length for a %s
Write the content now:`, contentType, topic, keywords, tone, contentType)

	reqBody := chatRequest{
		Model:    "deepseek/deepseek-chat-v3-0324",
		Messages: []chatMessage{{Role: "user", Content: prompt}},
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", openRouterURL, bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	apiKey := os.Getenv("OPENROUTER_API_KEY")
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("API request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(body))
	}

	var chatResp chatResponse
	if err := json.Unmarshal(body, &chatResp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(chatResp.Choices) == 0 {
		return "", fmt.Errorf("no choices in API response")
	}

	return chatResp.Choices[0].Message.Content, nil
}
